core/src/main/java/org/apache/calcite/util/RuleMatchVisualizerHook.java

```diff
@@ -0,0 +1,155 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import org.apache.calcite.config.CalciteConnectionConfig;
+import org.apache.calcite.jdbc.CalciteConnection;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.hep.HepPlanner;
+import org.apache.calcite.plan.visualizer.RuleMatchVisualizer;
+import org.apache.calcite.runtime.Hook;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+/**
+ * Utility class to enable RuleMatchVisualizer for Calcite connections.
+ *
+ * <p>This class provides hooks to automatically attach a RuleMatchVisualizer
+ * to planners when a connection specifies the ruleVisualizerDir property.
+ *
+ * <p>Usage in JDBC URL:
+ * <blockquote><pre>
+ * jdbc:calcite:ruleVisualizerDir=/tmp/calcite-viz
+ * </pre></blockquote>
+ *
+ * <p>Or programmatically:
+ * <blockquote><pre>
+ * RuleMatchVisualizerHook.enable("/tmp/calcite-viz");
+ * </pre></blockquote>
+ */
+public class RuleMatchVisualizerHook {
+  public static final RuleMatchVisualizerHook INSTANCE = new RuleMatchVisualizerHook();
+
+  private final Map<RelOptPlanner, RuleMatchVisualizer> visualizerMap = new HashMap<>();
+  private final AtomicInteger queryCounter = new AtomicInteger(0);
+
+  private Hook.Closeable hookCloseable = Hook.Closeable.EMPTY;
+
+  /** Private constructor to prevent instantiation. */
+  private RuleMatchVisualizerHook() {}
+
+  /**
+   * Enables the visualizer for all subsequent queries with the specified output directory.
+   *
+   * @param outputDir Directory where visualization files will be created
+   */
+  public synchronized void enable(String outputDir) {
+    hookCloseable.close();
+
+    // Ensure the output directory exists
+    File dir = new File(outputDir);
+    if (!dir.exists()) {
+      boolean madeDir = dir.mkdirs();
+      assert madeDir : "Failed to create directory: " + outputDir;
+    }
+
+    // Install the hook
+    hookCloseable = Hook.PLANNER.addThread((Consumer<RelOptPlanner>) planner -> {
+      attachVisualizer(planner, outputDir);
+    });
+  }
+
+  /**
+   * Enables the visualizer using the connection's configuration.
+   * This method checks if the connection has the ruleVisualizerDir property set.
+   *
+   * @param connection The Calcite connection
+   */
+  public synchronized void enableFromConnection(CalciteConnection connection) {
```

Comment:
The code in [CalciteConnectionImpl](https://github.com/apache/calcite/pull/4602/files#diff-312ec137d81d2e343908c9c879bbf31a51f363df2e1e633f764aa6702715987eR198) already checks for an existing dir, so it could just call the `enable` method directly.